### PR TITLE
Calibration rvc3

### PR DESCRIPTION
**Pull request by MaticTonin**
Update on calibrate.py file.

### CODE REVIEW

File: calibrate.py [hunk 1]
```diff
@@ -87,20 +153,39 @@ def parse_args():
                         help="Invert vertical axis of the camera for the display")
     parser.add_argument("-ih", "--invertHorizontal", dest="invert_h", default=False, action="store_true",
                         help="Invert horizontal axis of the camera for the display")
-    parser.add_argument("-ep", "--maxEpiploarError", default="1.0", type=float, required=False,
-                        help="Sets the maximum epiploar allowed with rectification")
+    # parser.add_argument("-ep", "--maxEpiploarError", default="1.0", type=float, required=False,
+    #                     help="Sets the maximum epiploar allowed with rectification")
```
Comment (saching13) [inline calibrate.py:157]: Bring this back. and set the default to 0.7

File: calibrate.py [hunk 2]
```diff
@@ -87,20 +153,39 @@ def parse_args():
...
+                        help="Path to dataset used for processing images")
+    parser.add_argument('-mdmp', '--minDetectedMarkersPercent', type=float, default=0.5,
+                        help="Minimum percentage of detected markers to consider a frame valid")
+    parser.add_argument('-nm', '--numMarkers', type=int, default=None, help="Number of markers in the board")
+    parser.add_argument('-mt', '--mouseTrigger', default=False, action="store_true",
+                        help="Enable mouse trigger for image capture")
+    parser.add_argument('-nic', '--noInitCalibration', default=False, action="store_true",
+                        help="Don't take the board calibration for initialization but start with an empty one")
+    parser.add_argument('-dismin', '--displayMinimum', default=False, action="store_true",
+                        help="Display the new minimum every second while taking pictures")
```
Comment (saching13) [inline calibrate.py:185]: what minimum is this @moratom

File: calibrate.py [hunk 3]
```diff
@@ -114,6 +199,186 @@ def parse_args():
...
+                    # Retrieve and pop the others
+                    synced = {}
+                    for name in indicies.keys():
+                        synced[name] = self.queues[name][min_ts_diff['indicies'][name]]
+                        # pop out the older messages
+                        for i in range(0, min_ts_diff['indicies'][name]+1):
+                            self.queues[name].popleft()
+
+                    print('Returning synced messages with error:', min_ts_diff['ts'], min_ts_diff['indicies'])
+                    return synced
```
Comment (saching13) [inline calibrate.py:379]: make this part of `TraceLevel` or make it part of `displayMinimum` by replacing the `displayMinimum` name to `enableDebugMessageSync`

File: depthai_helpers/calibration_utils.py [hunk 4]
```diff
@@ -270,181 +293,209 @@ def analyze_charuco(self, images, scale_req=False, req_resolution=(800, 1280)):
...
+        image_files.sort()
+        assert len(
+            image_files) != 0, "ERROR: Images not read correctly, check directory"
+
+        allCorners, allIds, _, _, imsize, _ = self.analyze_charuco(image_files)
+        # print(imsize)
+        # coverageShape = (imsize[0], imsize[1], 3)
+        #print(imsize)
+        #print(imsize[-1])
+        #print(imsize[::-1])
```
Comment (saching13) [inline depthai_helpers/calibration_utils.py:321]: clean this up

File: depthai_helpers/calibration_utils.py [hunk 5]
```diff
@@ -270,181 +293,209 @@ def analyze_charuco(self, images, scale_req=False, req_resolution=(800, 1280)):
...
-
-        self.H1 = np.matmul(np.matmul(self.M2, self.R1),
-                            np.linalg.inv(self.M1))
-        self.H2 = np.matmul(np.matmul(self.M2, self.R2),
-                            np.linalg.inv(self.M2))
+            images_right) != 0, "ERROR: Images not found, check directory"
+        # print('Images from left and right')
+        # print(images_left[0])
+        # print(images_right[0])
+
```
Comment (saching13) [inline depthai_helpers/calibration_utils.py:366]: clean up

File: depthai_helpers/calibration_utils.py [hunk 6]
```diff
@@ -469,221 +522,339 @@ def calibrate_fisheye(self, allCorners, allIds, imsize):
...
         one_pts = self.board.chessboardCorners
-        print('allIds_l')
-        print(len(allIds_l))
-        print(len(allIds_r))
-        print('allIds_l')
+        # print('allIds_l')
+        # print(len(allIds_l))
+        # print('allIds_r')
+        # print(len(allIds_r))
+        # print('allIds_l')
```
Comment (saching13) [inline depthai_helpers/calibration_utils.py:552]: put them under TraceLevel

File: depthai_helpers/calibration_utils.py [hunk 7]
```diff
@@ -469,221 +522,339 @@ def calibrate_fisheye(self, allCorners, allIds, imsize):
...
+                print(R)
+                print(T)
+                print(f'Euler angles in XYZ {r_euler} degs')
+
+
+            # ret, M1, d1, M2, d2, R, T, E, F = cv2.stereoCalibrate(
+            #     obj_pts, left_corners_sampled, right_corners_sampled,
+            #     cameraMatrix_l, distCoeff_l, cameraMatrix_r, distCoeff_r, imsize,
+            #     criteria=stereocalib_criteria, flags=flags)
+            # if np.absolute(T[1])  > 0.2:
```
Comment (saching13) [inline depthai_helpers/calibration_utils.py:612]: cleanup

**Review by saching13 — CHANGES_REQUESTED** [review]
Added some fixes. 
- depthai_helpers/calibration_utils_old.py remove this file
- remove the local board files and make sure the code to use depthai_boards submule

File: calibrate.py (hunk 1 above)
Comment (MaticTonin) [reply calibrate.py:157]: Changed.

File: calibrate.py (hunk 3 above)
Comment (MaticTonin) [reply calibrate.py:379]: I changed displayMinimum into enableDebugMessageSync in all cases.

File: calibrate.py (hunk 2 above)
Comment (MaticTonin) [reply calibrate.py:185]: I used argument for display of messages. Now it is changed to enableDebugMessageSync.

**Review by saching13 — APPROVED** [review]
LGTM. 
Please test calibration on OAK-D, OAK-D-W. and a PoE device before merging and  check the visual depth quality before merging.

File: depthai_helpers/calibration_utils.py [hunk 8]
```diff
@@ -187,15 +204,32 @@ def calibrate(self, board_config, filepath, square_size, mrk_size, squaresX, squ
                                                                                         extrinsics[3], # Left Rectification rotation 
                                                                                         extrinsics[4], # Right Rectification rotation 
                                                                                         extrinsics[5], # Left Rectification Intrinsics
-                                                                                        extrinsics[6]) # Right Rectification Intrinsics
-
+                                                                                        extrinsics[6], # Right Rectification Intrinsics
+                                                                                        display_all)
```
Comment (saching13) [inline depthai_helpers/calibration_utils.py:208]: avoid lines getting filled with spaces

File: depthai_helpers/calibration_utils.py [hunk 9]
```diff
@@ -87,17 +87,18 @@ def polygon_from_image_name(image_name):
 
 
 class StereoCalibration(object):
+    global traceLevel
```
Comment (saching13) [inline depthai_helpers/calibration_utils.py:90]: initialize this trace level when you call the calibrate from calibrate.py

File: depthai_helpers/calibration_utils.py [hunk 10]
```diff
@@ -564,6 +750,23 @@ def display_rectification(self, image_data_pairs, images_corners_l, images_corne
...
+                #          (0, line_row), (img_concat.shape[1], line_row),
+                #          (0, 255, 0), 1)
+                # line_row += 30
+
+            # line_col = 0
+            # while not isHorizontal and line_col < img_concat.shape[1]:
+            #     cv2.line(img_concat,
+            #              (line_col, 0), (line_col, img_concat.shape[0]),
+            #              (0, 255, 0), 1)
+            #     line_col += 40
```
Comment (saching13) [inline depthai_helpers/calibration_utils.py:769]: remove this

File: depthai_helpers/calibration_utils.py [hunk 11]
```diff
@@ -574,9 +777,180 @@ def display_rectification(self, image_data_pairs, images_corners_l, images_corne
...
+            # while line_row < img_concat.shape[0]:
+            #     cv2.line(img_concat,
+            #              (0, line_row), (img_concat.shape[1], line_row),
+            #              (0, 255, 0), 1)
+            #     line_row += 30
+
+            # cv2.imshow('Stereo Pair', img_concat)
+            # k = cv2.waitKey(0)
+            # if k == 27:  # Esc key to stop
+            #     break
```
Comment (saching13) [inline depthai_helpers/calibration_utils.py:895]: clean up

**Review by saching13 — CHANGES_REQUESTED** [review]

File: calibrate.py [hunk 12]
```diff
@@ -27,38 +27,62 @@
...
+                'RIGHT' : dai.CameraBoardSocket.CAM_C,
+                'CAM_A' : dai.CameraBoardSocket.CAM_A,
+                'CAM_B' : dai.CameraBoardSocket.CAM_B,
+                'CAM_C' : dai.CameraBoardSocket.CAM_C,
+                'CAM_D' : dai.CameraBoardSocket.CAM_D,
+                'CAM_E' : dai.CameraBoardSocket.CAM_E,
+                'CAM_F' : dai.CameraBoardSocket.CAM_F,
+                'CAM_G' : dai.CameraBoardSocket.CAM_G,
+                'CAM_H' : dai.CameraBoardSocket.CAM_H
+                }
```
[the remark this answers is not in the document]
Comment (MaticTonin) [reply calibrate.py:49]: Cleaned

File: calibrate.py [hunk 13]
```diff
@@ -27,38 +27,62 @@
...
+                dai.CameraBoardSocket.CAM_C : 'RIGHT',
+                dai.CameraBoardSocket.CAM_A : 'CAM_A',
+                dai.CameraBoardSocket.CAM_B : 'CAM_B',
+                dai.CameraBoardSocket.CAM_C : 'CAM_C',
+                dai.CameraBoardSocket.CAM_D : 'CAM_D',
+                dai.CameraBoardSocket.CAM_E : 'CAM_E',
+                dai.CameraBoardSocket.CAM_F : 'CAM_F',
+                dai.CameraBoardSocket.CAM_G : 'CAM_G',
+                dai.CameraBoardSocket.CAM_H : 'CAM_H'
+                }
```
[the remark this answers is not in the document]
Comment (MaticTonin) [reply calibrate.py:67]: As I see it is not used.

File: calibrate.py [hunk 14]
```diff
@@ -199,15 +317,15 @@ def add_msg(self, name, msg):
         #     print()
         # print()
 
-    def get_synced(self):
+    def get_synced(self, enableDebugMessageSync):
```
[the remark this answers is not in the document]
Comment (MaticTonin) [reply calibrate.py:320]: Done in self.enableDebugMessageSync.

File: depthai_helpers/calibration_utils.py (hunk 10 above)
Comment (MaticTonin) [reply depthai_helpers/calibration_utils.py:769]: Cleaned

File: depthai_helpers/calibration_utils.py (hunk 11 above)
Comment (MaticTonin) [reply depthai_helpers/calibration_utils.py:895]: Cleaned

File: calibrate.py [hunk 15]
```diff
@@ -914,265 +1067,229 @@ def capture_images(self):
...
+                    calibHandler = self.device.readCalibration2()
+                    eepromUnionData['calibrationUser'] = calibHandler.eepromToJson()
+
+                    calibHandler = self.device.readFactoryCalibration()
+                    eepromUnionData['calibrationFactory'] = calibHandler.eepromToJson()
+
+                    eepromUnionData['calibrationUserRaw'] = self.device.readCalibrationRaw()
+                    eepromUnionData['calibrationFactoryRaw'] = self.device.readFactoryCalibrationRaw()
+                    with open(calib_dest_path, "w") as outfile:
+                        json.dump(eepromUnionData, outfile, indent=4) """
```
[the remark this answers is not in the document]
Comment (MaticTonin) [reply calibrate.py:1236]: Cleaned

File: calibrate.py [hunk 16]
```diff
@@ -741,7 +903,6 @@ def capture_images(self):
         prev_time = None
         curr_time = None
         self.display_name = "left + right + rgb"
-        last_frame_time = time.time()
         # with self.get_pipeline() as pipeline:
         while not finished:
             current_left  = self.left_camera_queue.tryGet()
```
[the remark this answers is not in the document]
Comment (MaticTonin) [reply calibrate.py:908]: Cleaned

File: calibrate.py [hunk 17]
```diff
@@ -302,66 +408,116 @@ def __init__(self):
...
             if properties.sensorName == 'OV7251':
                 raise Exception(
-            "OAK-D-Lite Calibration is not supported on main yet. Please use `lite_calibration` branch to calibrate your OAK-D-Lite!!")
+            "OAK-D-Lite Calibration is not supported on main yet. Please use `lite_calibration` branch to calibrate your OAK-D-Lite!!") 
 
         self.device.startPipeline(pipeline)"""
         # self.left_camera_queue = self.device.getOutputQueue("left", 30, True)
         # self.right_camera_queue = self.device.getOutputQueue("right", 30, True)
         # if not self.args.disableRgb:
         #     self.rgb_camera_queue = self.device.getOutputQueue("rgb", 30, True)
```
[the remark this answers is not in the document]
Comment (MaticTonin) [reply calibrate.py:464]: Cleaned

File: calibrate.py [hunk 18]
```diff
@@ -302,66 +408,116 @@ def __init__(self):
...
-                cam = self.board_config['cameras'][config_cam]
-                self.camera_queue[cam['name']] = self.device.getOutputQueue(cam['name'], 1, False)
+        
 
         """ cameraProperties = self.device.getConnectedCameraProperties()
         for properties in cameraProperties:
             if properties.sensorName == 'OV7251':
                 raise Exception(
-            "OAK-D-Lite Calibration is not supported on main yet. Please use `lite_calibration` branch to calibrate your OAK-D-Lite!!")
+            "OAK-D-Lite Calibration is not supported on main yet. Please use `lite_calibration` branch to calibrate your OAK-D-Lite!!")
```
[the remark this answers is not in the document]
Comment (MaticTonin) [reply calibrate.py:458]: Cleaned

File: DepthCapture_2023-06-27-14-50-40.png
Comment (saching13) [inline DepthCapture_2023-06-27-14-50-40.png:1]: where did this come from ?

File: calibrate.py [hunk 19]
```diff
@@ -167,16 +171,107 @@ def parse_args():
...
+            self.recentFrameTs = ts
+        # print(len(self.arrays[name]))
+        # print(f'Added Msgs typ {name}')
+        # print(ts)
+        # for name, arr in self.arrays.items():
+        #     for i, obj in enumerate(arr):
+        #         if self.remove(obj['timestamp']):
+        #             arr.remove(obj)
+        #         else: break
+
```
Comment (saching13) [inline calibrate.py:203]: clean up

File: DepthCamera_2023-06-27-14-50-40.json
Comment (saching13) [inline DepthCamera_2023-06-27-14-50-40.json:1]: what is this for ?

File: resources/boards/veermer_no_rgb.json
Comment (saching13) [inline resources/boards/veermer_no_rgb.json:1]: remove this

**Review by saching13 — COMMENTED** [review]
added additional comments

File: calibrate.py (hunk 2 above)
Comment (saching13) [reply calibrate.py]: make it something like `debugMinoffsetSync` add better help explaination

File: depthai_helpers/calibration_utils.py [hunk 20]
```diff
@@ -89,15 +88,15 @@ def polygon_from_image_name(image_name):
 class StereoCalibration(object):
     """Class to Calculate Calibration and Rectify a Stereo Camera."""
 
-    def __init__(self, traceLevel: int):
-        self.traceLevel = traceLevel
+    def __init__(self):
         """Class to Calculate Calibration and Rectify a Stereo Camera."""
 
-    def calibrate(self, board_config, filepath, square_size, mrk_size, squaresX, squaresY, camera_model, enable_disp_rectify):
+    def calibrate(self, board_config, filepath, square_size, mrk_size, squaresX, squaresY, camera_model, enable_disp_rectify, display_all):
```
Comment (saching13) [inline depthai_helpers/calibration_utils.py:94]: What is display_all ?

I would rather have traceLevel with some maxNumber which does displayall.

Issue with display all is it floods your terminal making it hard to debug.

File: depthai_helpers/calibration_utils.py [hunk 21]
```diff
@@ -209,16 +241,38 @@ def analyze_charuco(self, images, resize_img_func = None):
         # SUB PIXEL CORNER DETECTION CRITERION
         criteria = (cv2.TERM_CRITERIA_EPS +
                     cv2.TERM_CRITERIA_MAX_ITER, 10000, 0.00001)
+        count = 0
         for im in images:
             if self.traceLevel == 2:
```
Comment (saching13) [inline depthai_helpers/calibration_utils.py:246]: Where is the traceLevel initialized. you seem to have removed it and never initialized it from the function Class level. If ran. this code might break. 
Did you check the execution ?

**Review by saching13 — CHANGES_REQUESTED** [review]
Please fix the trace level operation and don't introduce display_all.

File: DepthCapture_2023-06-27-14-50-40.png
Comment (MaticTonin) [reply DepthCapture_2023-06-27-14-50-40.png:1]: Cleared.

File: calibrate.py (hunk 19 above)
Comment (MaticTonin) [reply calibrate.py:203]: Cleared.

File: DepthCamera_2023-06-27-14-50-40.json
Comment (MaticTonin) [reply DepthCamera_2023-06-27-14-50-40.json:1]: Cleared.

File: resources/boards/veermer_no_rgb.json
Comment (MaticTonin) [reply resources/boards/veermer_no_rgb.json:1]: Cleared.

File: depthai_helpers/calibration_utils.py (hunk 20 above)
Comment (MaticTonin) [reply depthai_helpers/calibration_utils.py:94]: Setting the trace value to 10, if you want to display all the data in calibration.

File: depthai_helpers/calibration_utils.py (hunk 21 above)
Comment (MaticTonin) [reply depthai_helpers/calibration_utils.py:246]: It is initialized in calibrate.py as an argument and then passed to stereo_calib in line 824 in calibrate.py.
I checked on OAK-D-PRO and it worked fine.

**Review by saching13 — APPROVED** [review]

File: calibrate.py [hunk 22]
```diff
@@ -144,19 +143,24 @@ def parse_args():
...
     parser.add_argument('-mt', '--mouseTrigger', default=False, action="store_true",
                         help="Enable mouse trigger for image capture")
-    parser.add_argument('-l', '--traceLevel', type=int, default=2,
-                        help="Set the debug trace level. Default: %(default)s.")
-    
+    parser.add_argument('-nic', '--noInitCalibration', default=False, action="store_true",
+                        help="Don't take the board calibration for initialization but start with an empty one")
+    parser.add_argument('-trc', '--traceLevel', type=int, default=0,
+                        help="Set to trace the steps in calibration. Number from 1 to 5. If you want to display all, set trace number to 10.")
+    parser.add_argument('-disall', '--enableDebugMessageSync', default=False, action="store_true",
```
[the remark this answers is not in the document]
Comment (saching13) [reply calibrate.py]: Fixed above.

**Pull request merged**